From 1081b8f07bc929a52aeb2d005105d5550c293591 Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Thu, 10 Sep 2026 14:12:49 +0200
Subject: [PATCH] alsa-plugin: wake up poll on errors and stop

Also wake up the poll fd when we have an xrun or if we are in the
SETUP state. This is because the revents should report POLLERR in
those error cases.

Force wakeup when stopping because this could be because the state is
changing to SETUP and we need to report POLLERR.

Fixes #5444

diff --git a/pipewire-alsa/alsa-plugins/pcm_pipewire.c b/pipewire-alsa/alsa-plugins/pcm_pipewire.c
index 73b00446e..9399e7459 100644
--- a/pipewire-alsa/alsa-plugins/pcm_pipewire.c
+++ b/pipewire-alsa/alsa-plugins/pcm_pipewire.c
@@ -87,7 +87,7 @@ typedef struct {
 
 static int snd_pcm_pipewire_stop(snd_pcm_ioplug_t *io);
 
-static int update_active(snd_pcm_ioplug_t *io)
+static int update_active(snd_pcm_ioplug_t *io, bool force_wakeup)
 {
 	snd_pcm_pipewire_t *pw = io->private_data;
 	snd_pcm_sframes_t avail;
@@ -96,7 +96,9 @@ static int update_active(snd_pcm_ioplug_t *io)
 
 	avail = snd_pcm_ioplug_avail(io, pw->hw_ptr, io->appl_ptr);
 
-	if (pw->error > 0) {
+	if (pw->error > 0 || force_wakeup ||
+	    io->state == SND_PCM_STATE_SETUP ||
+	    pw->xrun_detected) {
 		active = true;
 	}
 	else if (io->state == SND_PCM_STATE_DRAINING) {
@@ -167,8 +169,10 @@ static int snd_pcm_pipewire_poll_revents(snd_pcm_ioplug_t *io,
 		return pw->error;
 
 	*revents = pfds[0].revents & ~(POLLIN | POLLOUT);
-	if (pfds[0].revents & POLLIN && update_active(io))
+	if (pfds[0].revents & POLLIN && update_active(io, false))
 		*revents |= (io->stream == SND_PCM_STREAM_PLAYBACK) ? POLLOUT : POLLIN;
+	if (io->state == SND_PCM_STATE_SETUP || pw->xrun_detected)
+		*revents |= POLLERR;
 
 	pw_log_trace_fp("poll %d", *revents);
 
@@ -392,7 +396,7 @@ static void on_stream_state_changed(void *data, enum pw_stream_state old, enum p
 	if (state == PW_STREAM_STATE_ERROR) {
 		pw_log_warn("%s", error);
 		pw->error = -errno;
-		update_active(&pw->io);
+		update_active(&pw->io, false);
 	}
 }
 
@@ -466,7 +470,7 @@ static void on_stream_process(void *data)
 		}
 	}
 done:
-	update_active(io);
+	update_active(io, false);
 }
 
 static const struct pw_stream_events stream_events = {
@@ -611,7 +615,7 @@ static int snd_pcm_pipewire_stop(snd_pcm_ioplug_t *io)
 	snd_pcm_pipewire_t *pw = io->private_data;
 
 	pw_log_debug("%p: stop", pw);
-	update_active(io);
+	update_active(io, true);
 
 	pw_thread_loop_lock(pw->main_loop);
 	if (pw->activated && pw->stream != NULL) {
@@ -1227,7 +1231,7 @@ static void on_core_error(void *data, uint32_t id, int seq, int res, const char
 	if (id == PW_ID_CORE) {
 		pw->error = res;
 		if (pw->fd != -1)
-			update_active(&pw->io);
+			update_active(&pw->io, false);
 	}
 	pw_thread_loop_signal(pw->main_loop, false);
 }

